### PR TITLE
Filter wildcards using searchResults

### PR DESCRIPTION
**Pull request by KyleJu**
Fix https://github.com/web-platform-tests/wpt.fyi/issues/2063

### CODE REVIEW

**Comment by KyleJu** [comment]
Deployment is working and will add some tests.

**Comment by Hexcles** [comment]
`PathInfo.getDirname` is not a function?

**Comment by KyleJu** [comment]
> `PathInfo.getDirname` is not a function?

its still outdated. I have fixed that issue in the latest commit `ad828d6`

**Comment by Hexcles** [comment]
I wonder if something's wrong with draft PRs...

**Comment by KyleJu** [comment]
> I wonder if something's wrong with draft PRs...

I add some tests and push another commit. Hope it fixes the issue

File: webapp/components/wpt-metadata.js [hunk 1]
```diff
@@ -267,7 +276,8 @@ class WPTMetadata extends PathInfo(LoadingState(PolymerElement)) {
         curPath = curPath + '/';
       }
       const dirname = testname.substring(0, testname.length - 1);
-      return curPath.startsWith(dirname) || this.isParentDir(curPath, dirname);
+      const dirnameWithoutSlash = testname.substring(0, testname.length - 2);
+      return curPath.startsWith(dirname) || (this.isParentDir(curPath, dirname) && testResultSet.has(dirnameWithoutSlash));
```
Comment (Hexcles) [inline webapp/components/wpt-metadata.js:280]: Is there a way to always use one of `dirname` and `dirnameWithoutSlash`? IIRC `isParentDir` accepts both?

File: webapp/components/wpt-metadata.js [hunk 2]
```diff
@@ -192,7 +192,16 @@ class WPTMetadata extends PathInfo(LoadingState(PolymerElement)) {
       return;
     }
 
-    return new Set(searchResults.map(r => r.test));
+    const testResultSet = new Set();
+    for (const result of searchResults) {
+      var test = result.test;
+      // Dir terminates with ''
```
[the remark this answers is not in the document]
Comment (KyleJu) [reply webapp/components/wpt-metadata.js]: Oops I clarified the comment. Most of getDirname APIs terminates with `/` so I thought I would leave a comment here.

File: webapp/components/wpt-metadata.js (hunk 1 above)
Comment (KyleJu) [reply webapp/components/wpt-metadata.js:280]: Correct, `isParentDir` accepts both. However, if using `dirnameWithoutSlash` for all of them, there is an edge case when `curpath` is `/abcd.html and `dirname` is `/abc`; using `dirname` for `testResultSet.has(dirnameWithoutSlash)` makes the implementation of `getDirname` more complicated. I thought it wasn't worth the work there

Comment (Hexcles) [reply webapp/components/wpt-metadata.js:280]: But you could add `test + '/'` to `testResultSet` without having to change `getDirname`?

Comment (KyleJu) [reply webapp/components/wpt-metadata.js:280]: Right but testResultSet has both directories and test name. To do `test + '/'`, we need to add a check when test is a testname where we don't want to add `/` (or `testResultSet.add(test)` before going into the loop) . Same goes for cases when test is `/`.  These make line 200 and 201 less elegant. But I am open to it

Comment (Hexcles) [reply webapp/components/wpt-metadata.js:280]: Could you remind me what the purpose of `curPath.startsWith(dirname)` is? If it's not necessary, can we just keep the second part and always use `dirnameWithoutSlash`?

Comment (KyleJu) [reply webapp/components/wpt-metadata.js:280]: The purpose is that a metadata with wildcard shows up in all children directories and I do think it is necessary

**Comment by KyleJu** [comment]
@Hexcles friendly ping

File: webapp/components/wpt-metadata.js (hunk 1 above)
Comment (Hexcles) [reply webapp/components/wpt-metadata.js:280]: Reading this change again, I wonder if the condition should be

```suggestion
      return (curPath.startsWith(dirname) || this.isParentDir(curPath, dirname) && testResultSet.has(dirnameWithoutSlash);
```

And we should probably make the names more meaningful (e.g. `testname` -> `metadataTestName`, `checkPath` -> `shouldShowMetadata`) and add some comments to each clause of this boolean expression.

**Comment by Hexcles** [comment]
There's also a merge conflict.

File: webapp/components/wpt-metadata.js (hunk 1 above)
Comment (KyleJu) [reply webapp/components/wpt-metadata.js:281]: Sound good. The fact that you are able to see `curPath` implies `testResultSet.has(dirnameWithoutSlash)` is true but this is more rigid nevertheless

**Comment by KyleJu** [comment]
@Hexcles conflict resolved but I feel like navigating in [this wpt.fyi staging](https://filter-results-for-wil-dot-wptdashboard-staging.uk.r.appspot.com/results/?label=master&label=experimental&aligned) is visibly slower? Or its already slow

**Comment by Hexcles** [comment]
@KyleJu got a specific example (maybe a search?) where this staging deployment is slower?

**Comment by KyleJu** [comment]
> @KyleJu got a specific example (maybe a search?) where this staging deployment is slower?

e.g. clicking `css-fonts/` [on this staging](https://filter-results-for-wil-dot-wptdashboard-staging.uk.r.appspot.com/results/css?label=experimental&label=master&aligned) vs [wpt.fyi ](https://wpt.fyi/results/css?label=experimental&label=master&aligned) without enabling `Show metadata Information on the wpt.fyi result page.` I feel like the former is a second slower

**Comment by Hexcles** [comment]
I tried capturing some performance profiles.

wpt.fyi without showing metadata took about 300ms while both wpt.fyi and this staging instance with showing metadata took about 500ms for this navigation, so I don't think this change is causing much performance regression, but rather showing the metadata is a bit slow.

**Comment by KyleJu** [comment]
GitHub checks still failing after rerunning 3 times. https://github.com/web-platform-tests/wpt.fyi/issues/2076

**Comment by Hexcles** [comment]
admin-merged

**Pull request merged**